src/mds/Server.h

```diff
@@ -91,6 +93,7 @@ class Server {
 
 public:
   bool terminating_sessions;
+  set<int> crs_msg_seq_nums;
```

Comment:
This should actually be a pair of sequence numbers (`version_t`) that are associated with each session (just like seq numbers in TCP). The first would be what the MDS sends and sequentially increases with each new `CEPH_SESSION_RECALL_STATE` message. The second would be what the client sends back as finished, via the `CEPH_SESSION_RECALL_STATEMSG_ACK` message.

But now that I'm reading the code more in SessionMap.h, I'm wondering if we can reuse the existing `CEPH_SESSION_FLUSHMSG` code for this purpose. There is already a sequence number in the session to use (see the code for `Session::cap_push_seq`. @ukernel would this be appropriate or should we use a separate sequence number?

Reply:
using Session::cap_push_seq is appropriate. 

If client always sends ACK immediately after receiving drop cache request, existing Session::{wait_for_flush,finish_flush} code can serve this purpose completely. If not, need introduce a new waiter list ( similar to Session::waitfor_flush)

Reply:
@batrick 

> This should actually be a pair of sequence numbers (version_t) that are associated with each session (just like seq numbers in TCP). The first would be what the MDS sends and sequentially increases with each new CEPH_SESSION_RECALL_STATE message. The second would be what the client sends back as finished, via the CEPH_SESSION_RECALL_STATEMSG_ACK message.

If sequence number of the message sent by MDS is different from that of the client, I wouldn't know which sessions have not been trimmed and it will not be possible to report those sessions in the log. Though I can still see how many sessions have not trimmed. Will that be okay?

Reply:
check code that handle CEPH_SESSION_FLUSHMSG/CEPH_SESSION_FLUSHMSG_ACK. they use the same seq number

Reply:
@ukernel Done. Thanks.  :)